File: lego-osek-master/crv/systems/lego-osek/properties/edc_on_differential/EdcOnDifferentialMon.hs
module Main where

import Language.Copilot
import Copilot.Compile.C99
import Data.Int (Int8, Int32)

import Prelude hiding ((&&), (||), not, (==), (/=), (<=), (>=), (>), (<))

fr_edc_flag :: Stream Int8
fr_edc_flag = extern "fr_edc_flag" Nothing

fr_steering_angle :: Stream Int32
fr_steering_angle = extern "fr_steering_angle" Nothing

fr_cmd_speed :: Stream Int32
fr_cmd_speed = extern "fr_cmd_speed" Nothing

fr_motor_left :: Stream Int32
fr_motor_left = extern "fr_motor_left" Nothing

fr_motor_right :: Stream Int32
fr_motor_right = extern "fr_motor_right" Nothing

spec :: Spec
spec = do
  let edcOn = fr_edc_flag == (-1)
  let turning = (fr_steering_angle > 2) || (fr_steering_angle < (-2))
  let motorsDiffer = fr_motor_left /= fr_motor_right
  let p = (not (edcOn && turning)) || (fr_cmd_speed == 0) || motorsDiffer
  trigger "edc_on_differential_violation" (not p) []

main :: IO ()
main = reify spec >>= compile "edc_on_differential_mon"

File: lego-osek-master/crv/systems/lego-osek/properties/edc_on_differential/edc_on_differential_mon.c
#include <stdint.h>
#include <stdbool.h>
#include <string.h>
#include <stdlib.h>
#include <math.h>

#include "edc_on_differential_mon_types.h"
#include "edc_on_differential_mon.h"

static int8_t fr_edc_flag_cpy;
static int32_t fr_steering_angle_cpy;
static int32_t fr_cmd_speed_cpy;
static int32_t fr_motor_left_cpy;
static int32_t fr_motor_right_cpy;

static bool edc_on_differential_violation_0_guard(void) {
  return !((!(((fr_edc_flag_cpy) == ((int8_t)(-1))) && (((fr_steering_angle_cpy) > ((int32_t)(2))) || ((fr_steering_angle_cpy) < ((int32_t)(-2)))))) || (((fr_cmd_speed_cpy) == ((int32_t)(0))) || ((fr_motor_left_cpy) != (fr_motor_right_cpy))));
}

void step(void) {
  (fr_edc_flag_cpy) = (fr_edc_flag);
  (fr_steering_angle_cpy) = (fr_steering_angle);
  (fr_cmd_speed_cpy) = (fr_cmd_speed);
  (fr_motor_left_cpy) = (fr_motor_left);
  (fr_motor_right_cpy) = (fr_motor_right);
  if ((edc_on_differential_violation_0_guard)()) {
    {(edc_on_differential_violation)();}
  };
}
